import code in images (OCR)

hatemhosny opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.

  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the LiveCodes Contributing Guide: https://github.com/live-codes/livecodes/blob/HEAD/CONTRIBUTING.md.
  - 📖 Read the LiveCodes Code of Conduct: https://github.com/live-codes/livecodes/blob/HEAD/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
  - 🌐 Use `window.deps.translateString` in .ts files and add `data-i18n` attributes in .html files to mark strings that needs to be translated.
-->

## What type of PR is this? (check all applicable)

- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] ♻️ Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert
- [ ] 🌐 Internationalization / Translation

## Description

This PR allows importing code in images (local or via URL).

[Tesseract.js](https://github.com/naptha/tesseract.js) is used for client-side OCR.

Language detection is performed using [highlight.js](https://highlightjs.readthedocs.io/en/latest/api.html#highlightauto)

Best results are obtained when the image is generated using LiveCodes "[Code to Image](https://livecodes.io/docs/features/code-to-image)" feature, with a share URL added to the image.

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentations?

- [x] 📓 docs (./docs)
- [ ] 📕 storybook (./storybook)
- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

hatemhosny commented:
@BassemHalim 

> I think removing numbers from the start of each line and reformatting would be enough

Line numbers are now removed.

However, formatting would be difficult!
Prettier and most formatter will not format if there are any syntax errors.
So any error with OCR will prevent formatting.

[Biome](https://biomejs.dev/) is a formatter that is compatible with Prettier, but would format blocks with no errors and ignore those with errors. But it is still lagging behind in [language support](https://biomejs.dev/internals/language-support/).
Maybe we can plan to move to Biome later.
I think for now, we should just let the user fix the errors and then trigger the format manually.

hatemhosny commented:
> ./build/livecodes/import.js    |    16.7 kB    |    +9.81 kB (+143.41%)    |    🆘

The [size check](https://github.com/live-codes/livecodes/pull/837#issuecomment-2923347084) shows significant increase in the `import` chunk bundle size.

This is related to this issue: #840 
I will just keep this as it is for now and then fix it with that planned refactor.

zyf722 commented:
Hi @hatemhosny,

Thanks for the opportunity to review this PR - looks like another great feature!

I have a few comments and questions:

1. When a user first tries to import from code, a `.traineddata` file is downloaded but during which there's no visual feedback in the UI. It would be nice to display a progress bar or some other indication that something is in progress in the background.

2. It seems that the detection quality for images with light theme is not as good as for dark theme. I tested this with the same code snippet rendered using both light and dark themes and had different results. For example:

  - [Code snippet with light theme](https://carbon.now.sh/?bg=rgba%28186%2C201%2C217%2C1%29&t=one-light&wt=none&l=auto&width=680&ds=true&dsyoff=20px&dsblur=68px&wc=true&wa=false&pv=56px&ph=56px&ln=true&fl=1&fm=Hack&fs=14px&lh=133%25&si=false&es=2x&wm=false&code=def%2520display_multilingual%28%29%253A%250A%2520%2520%2520%2520greetings%2520%253D%2520%255B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2522Hello%252C%2520World%21%2520%28English%29%2522%252C%250A%2520%2520%2520%2520%255D%250A%250A%2520%2520%2520%2520for%2520greeting%2520in%2520greetings%253A%250A%2520%2520%2520%2520%2520%2520%2520%2520print%28greeting%29%250A%250Aif%2520__name__%2520%253D%253D%2520%2522__main__%2522%253A%250A%2520%2520%2520%2520display_multilingual%28%29%250A)
    ![Light snippet](https://github.com/user-attachments/assets/079e5982-cbd5-4a33-8e22-0d18f51d0bd9)
    ![light](https://github.com/user-attachments/assets/3f7b5736-9ab1-435d-969c-0705aee40e78)
  - [Code snippet with dark theme](https://carbon.now.sh/?bg=rgba%28186%2C201%2C217%2C1%29&t=vscode&wt=none&l=auto&width=680&ds=true&dsyoff=20px&dsblur=68px&wc=true&wa=false&pv=56px&ph=56px&ln=true&fl=1&fm=Hack&fs=14px&lh=133%25&si=false&es=2x&wm=false&code=def%2520display_multilingual%28%29%253A%250A%2520%2520%2520%2520greetings%2520%253D%2520%255B%250A%2520%2520%2520%2520%2520%2520%2520%2520%2522Hello%252C%2520World%21%2520%28English%29%2522%252C%250A%2520%2520%2520%2520%255D%250A%250A%2520%2520%2520%2520for%2520greeting%2520in%2520greetings%253A%250A%2520%2520%2520%2520%2520%2520%2520%2520print%28greeting%29%250A%250Aif%2520__name__%2520%253D%253D%2520%2522__main__%2522%253A%250A%2520%2520%2520%2520display_multilingual%28%29%250A)
    ![Dark snippet](https://github.com/user-attachments/assets/7a7efc1e-9f47-40c5-a2e3-cf28a7d537e4)
    ![dark](https://github.com/user-attachments/assets/82d0b532-0865-44d9-944b-17e9ded8fc82)

3. I noticed the current implementation relies on detecting a LiveCodes share URL pattern in the last lines of the OCR'd text to extract the project ID.

   Since PNG images are being generated, would it be feasible to embed the project ID directly into the PNG metadata (e.g., using a [custom text chunk](https://www.libpng.org/pub/png/spec/iso/index-object.html#11textinfo))? This ID could then be read back during import. This approach could potentially make the import process more robust by removing the dependency on OCR accuracy for the URL pattern. Not sure about the implementation effort, but it might be an interesting alternative to consider.

zyf722 commented on src/livecodes/import/image.ts:
```diff
@@ -0,0 +1,106 @@
+  const lastLines = lines.slice(-2).join('\n');
+
+  // detect images created by LiveCodes "Code to Image" with share URL
+  const shareUrlPattern = /\?x=(id\/\S{11,20})/g;
```
To make the detection more accurate, would it be beneficial to check for a base URL or similar prefix here?

hatemhosny replied:
I'm not sure.
I want to support:
- hosted app: livecodes.io
- permanent URLs: v46.livecodes.io
- preview URLs: import-image.livecodes.pages.dev
- self-hosted instances: live-codes.github.io/livecodes (or any self-hosted URL)
- localhost:8080

Do you have a better suggestion?

hatemhosny replied:
Anyway, this should be a lot less relevant after using png meta tags.

hatemhosny commented:
hey @zyf722 
Great suggestions as usual!
Thank you.

> When a user first tries to import from code, a .traineddata file is downloaded but during which there's no visual feedback in the UI. It would be nice to display a progress bar or some other indication that something is in progress in the background.

You are right. I added a loading notification.

> It seems that the detection quality for images with light theme is not as good as for dark theme. I tested this with the same code snippet rendered using both light and dark themes and had different results.

This is the OCR quality of [tesseract.js](https://github.com/naptha/tesseract.js). I'm not sure we can improve that.
This is the best OCR library I found that works client-side.

> Since PNG images are being generated, would it be feasible to embed the project ID directly into the PNG metadata (e.g., using a [custom text chunk](https://www.libpng.org/pub/png/spec/iso/index-object.html#11textinfo))? This ID could then be read back during import. This approach could potentially make the import process more robust by removing the dependency on OCR accuracy for the URL pattern. Not sure about the implementation effort, but it might be an interesting alternative to consider.

That's such a nice suggestion. I found [meta-png](https://github.com/lucach/meta-png) which is a tiny library that reads and writes PNG meta data.
I implemented that and it had the following benefits:
- significant performance improvement.
- much more accurate.
- works even when share URL is not added to the image.

I looked for something similar for JPG. I only found [exiftool-vendored](https://www.npmjs.com/package/exiftool-vendored) which is a much larger library and I'm not even sure I'll be able to get it work in the browser.
So, I'll just do that for PNG (which is the default format).

So you can now generate images using the "Code-to-Image" tool and then try importing them.

Thanks again.

P.S. Importing from CodePen now works again (only pens of PRO accounts):
https://import-image.livecodes.pages.dev/?x=https://codepen.io/jkantner/pen/KKYZrRv

hatemhosny commented:
> Ah, I forgot there were three target formats for code-to-image. Then perhaps we can do the same for SVGs using the `<metadata>` tag, as browsers support it natively and no additional libraries are required?

I don't think we should deal with SVG as images in this sense.
If a user tries to import SVG, I expect he wants to edit the code.
This is how vs code and other editors do. I have personally edited SVG code in LiveCodes many times before.

I added a commit to handle importing SVG -> opens in html editor.
Also added a fallback: if an image has no text, it is added as an image tag in html (src is base64 data url)

hatemhosny commented:
.i18n-update-pull
